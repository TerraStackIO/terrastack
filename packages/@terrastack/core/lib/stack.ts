import { Construct, ISynthesisSession } from '@aws-cdk/core';
import * as fs from 'fs';
import * as path from 'path';
import { ResourceObject } from './resource-object';
import { resolve } from './_tokens';
import { removeEmpty } from './_util';
import { Names } from './names';
import {  
  snakeCase,
} from "./_util";


export class Stack extends Construct {

  /**
   * Finds the stack in which a node is defined.
   * @param node a construct node
   */
  public static of(node: Construct): Stack {
    if (node instanceof Stack) {
      return node;
    }

    const parent = node.node.scope as Construct;
    if (!parent) {
      throw new Error(`cannot find a parent chart (directly or indirectly)`);
    }

    return Stack.of(parent);
  }

  /**
   * The name of the stack's YAML file as emitted into the cloud assembly
   * directory during synthesis.
   */
  public readonly manifestFile: string;

  constructor(scope: Construct, ns: string) {
    super(scope, ns);
    this.manifestFile = `${this.node.uniqueId}.tf.json`;
  }

  /**
   * Generates a app-unique name for an object given it's construct node path.
   *
   * @param resourceObject The API object to generate a name for.
   */
  public generateObjectName(resourceObject: ResourceObject) {
    return Names.toDnsLabel(resourceObject.node.path);
  }

  protected synthesize(session: ISynthesisSession) {
    const doc: {[k: string]: {[k: string]: any}} = {};

    for (const resource of this.node.findAll()) {
      if (!(resource instanceof ResourceObject)) {
        continue;
      }
      const resourceName = snakeCase(resource.terraform.name)
      const type = snakeCase(resource.terraform.schemaType)
      

      if (!doc[type]) {
        const obj: {[k: string]: any} = {};
        doc[type] = obj
      }

      if (type === 'provider') {
        const manifest = removeEmpty(resolve(this, resource._render()));
        const merged = {...doc[type], ...manifest}
        doc[type] = merged
      } else {
        if (!doc[type][resourceName]) {
          const obj: {[k: string]: any} = {};
          doc[type][resourceName] = obj
        }
        
        const manifest = removeEmpty(resolve(this, resource._render()));
        const merged = {...doc[type][resourceName], ...manifest}
        doc[type][resourceName] = merged
      }
    }
    fs.writeFileSync(path.join(session.assembly.outdir, this.manifestFile), JSON.stringify(doc, null, 2));
  }
}