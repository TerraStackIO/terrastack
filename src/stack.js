/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

const path = require("path");
const ComponentProxy = require("./internals/component-proxy");
const _ = require("lodash");

class Stack {
  constructor(name, config) {
    this.name = name;
    this.config = config;
    this.components = [];
    this.executionOrder = [];
    this.terraStackDir = path.join(process.cwd(), ".terrastack", name);
  }

  add(...component) {
    this.components.push(
      ...component.map(c => new ComponentProxy(c, this.terraStackDir))
    );
  }

  resolve() {
    let availableComponents = this.components.slice(0);

    while (availableComponents.length > 0) {
      let countBefore = availableComponents.length;

      availableComponents = availableComponents.filter(component => {
        if (
          _.isEmpty(component.bindings) ||
          this._allBindingsAlreadyConsumed(component)
        ) {
          this.executionOrder.push(component);
          return false;
        } else {
          return true;
        }
      });

      // We expect the availableComponents array to change after each loop at least by one element.
      // So, if there was no change, some dependencies are circular or missing in the stack.
      if (availableComponents.length == countBefore) {
        throw "Circular or missing dependencies detected!";
      }
    }
    return this.executionOrder;
  }

  _allBindingsAlreadyConsumed(component) {
    return Object.values(component.bindings).every(binding => {
      if (this.executionOrder.includes(binding)) {
        return true;
      } else {
        return false;
      }
    });
  }
}

module.exports = Stack;
