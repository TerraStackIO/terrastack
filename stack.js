const path = require("path");
const fs = require("fs-extra");
const recursiveCopy = require("recursive-copy");
const _ = require("lodash");
const { spawnSync } = require("child_process");
const { Terraform } = require("./terraform");

class Stack {
  constructor(name, config) {
    this.name = name;
    this.config = config;
    this.components = [];
    this.executionOrder = [];
    this.terraStackDir = path.join(process.cwd(), ".terrastack", name);
  }

  add(...component) {
    this.components.push(...component);
  }

  async init(component, workingDir) {
    await run("init -force-copy", workingDir, { name: component.name });
  }

  async plan() {
    this.resolve();

    for (const component of this.executionOrder) {
      const callbacks = {
        start: () => {
          console.log(`${component.name} started`);
        },
        success: () => {
          console.log(`${component.name} success`);
        },
        failed: () => {
          console.log(`${component.name} failed`);
        }
      };
      const workingDir = await this._compile(component);
      const terraform = new Terraform(workingDir);

      await terraform.init(callbacks);

      await terraform.plan(
        Object.assign(callbacks, {
          success: changed => {
            if (changed)
              throw `${
                component.name
              } changed. Cannot proceed planing. Apply first.`;
          }
        })
      );

      await terraform.output(
        Object.assign(callbacks, {
          success: output => {
            component.outputs = this._unwrapOutputs(output);
          }
        })
      );
    }
  }

  async apply() {
    this.resolve();

    for (const component of this.executionOrder) {
      const callbacks = {
        start: () => {
          console.log(`${component.name} started`);
        },
        success: () => {
          console.log(`${component.name} success`);
        },
        failed: () => {
          console.log(`${component.name} failed`);
        }
      };
      const workingDir = await this._compile(component);
      const terraform = new Terraform(workingDir);

      await terraform.init(callbacks);

      await terraform.apply(callbacks);

      await terraform.output(
        Object.assign(callbacks, {
          success: output => {
            component.outputs = this._unwrapOutputs(output);
          }
        })
      );
    }
  }

  list() {
    this.resolve();
    return this.executionOrder;
  }

  async refresh(workingDir) {
    const obj = spawnSync(`terraform output -json`, {
      timeout: 0,
      shell: "/bin/bash",
      cwd: workingDir
    });

    return JSON.parse(obj.stdout);
  }

  async destroy() {
    this.resolve();

    for (const component of this.executionOrder.reverse()) {
      const callbacks = {
        start: () => {
          console.log(`${component.name} started`);
        },
        success: () => {
          console.log(`${component.name} success`);
        },
        failed: () => {
          console.log(`${component.name} failed`);
        }
      };
      const workingDir = await this._compile(component, true);
      const terraform = new Terraform(workingDir);

      await terraform.init(callbacks);

      await terraform.destroy(callbacks);
    }
  }

  resolve() {
    let availableComponents = this.components.slice(0);

    while (availableComponents.length > 0) {
      let countBefore = availableComponents.length;

      availableComponents = availableComponents.filter(component => {
        if (
          _.isEmpty(component.bindings) ||
          this._allBindingsAlreadyConsumed(component)
        ) {
          this.executionOrder.push(component);
          return false;
        } else {
          return true;
        }
      });

      // We expect the availableComponents array to change after each loop at least by one element.
      // So, if there was no change, some dependencies are circular or missing in the stack.
      if (availableComponents.length == countBefore) {
        throw "Circular or missing dependencies detected!";
      }
    }
  }

  async _compile(component, skipConfiguration = false) {
    fs.ensureDirSync(this.terraStackDir);

    const componentDir = path.join(this.terraStackDir, component.name);

    fs.ensureDirSync(componentDir);
    fs.writeFileSync(
      path.join(componentDir, "terrastack.tf"),
      JSON.stringify(this._compileConfig(component.name), null, 2)
    );

    if (!skipConfiguration) {
      fs.writeFileSync(
        path.join(componentDir, "terrastack.auto.tfvars"),
        JSON.stringify(component.optionsCallback(component.bindings), null, 2)
      );
    }

    await recursiveCopy(component.sourceDir, componentDir, {
      filter: ["**/*", "!.terrastack"],
      overwrite: true
    })
      .then(function(results) {
        console.info("Copied " + results.length + " files");
      })
      .catch(function(error) {
        console.error("Copy failed: " + error);
      });

    return componentDir;
  }

  _compileConfig(componentId) {
    const data = Object.keys(this.config).map(key =>
      this.config[key].compile(path.join(this.name, componentId))
    );

    return Object.assign({}, ...data);
  }

  _allBindingsAlreadyConsumed(component) {
    return Object.values(component.bindings).every(binding => {
      if (this.executionOrder.includes(binding)) {
        return true;
      } else {
        return false;
      }
    });
  }

  _unwrapOutputs(output) {
    return Object.entries(output).reduce((hash, element) => {
      const [key, value] = element;
      /* value is an object with the following keys:
       * sensitive: true/false
       * type: list/string
       * value: [...]/""
       */
      hash[key] = value["value"];
      return hash;
    }, {});
  }
}

module.exports = { Stack };
