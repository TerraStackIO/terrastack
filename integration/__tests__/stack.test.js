/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

const _ = require("lodash");
const fs = require("fs-extra");
const os = require("os");
const path = require("path");
const { runNode, addTerrastack } = require("../utils");

const runTerrastack = () => {
  const Terrastack = require("terrastack");
  const stack = require(process.cwd() + "/stack.js");
  new Terrastack(stack).init();
};

describe("Stack", () => {
  const rootDir = path.resolve(__dirname, "..", "..");
  const dir = path.resolve(rootDir, "examples", "simple");
  const workDir = path.resolve(os.tmpdir(), "simple");

  beforeEach(() => {
    fs.removeSync(path.resolve(workDir, "*"));
    fs.copySync(dir, workDir);
    fs.removeSync(path.resolve(workDir, "node_modules/*"));
    fs.removeSync(path.resolve(workDir, ".terrastack/*"));
    addTerrastack(workDir, rootDir);
  });

  it("applies a simple stack", () => {
    const apply = runNode(workDir, `${runTerrastack.toString()}()`);

    const state = runNode(
      workDir,
      `
        require(process.cwd() + "/stack.js").state("network")
      `
    );

    const result = _.slice(state.stdout.split("\n"), 2).sort();
    expect(result).toMatchSnapshot();
  });
});
