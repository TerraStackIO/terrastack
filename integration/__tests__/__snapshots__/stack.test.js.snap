// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack applies a simple stack 1`] = `
Array [
  "",
  "aws_internet_gateway.this",
  "aws_route.public_internet_gateway",
  "aws_route_table.private",
  "aws_route_table.public",
  "aws_route_table_association.private[0]",
  "aws_route_table_association.private[1]",
  "aws_route_table_association.private[2]",
  "aws_route_table_association.public[0]",
  "aws_route_table_association.public[1]",
  "aws_route_table_association.public[2]",
  "aws_subnet.private[0]",
  "aws_subnet.private[1]",
  "aws_subnet.private[2]",
  "aws_subnet.public[0]",
  "aws_subnet.public[1]",
  "aws_subnet.public[2]",
  "aws_vpc.this",
]
`;
